plutus-contract/src/Language/Plutus/Contract/Request.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE DerivingVia          #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE GADTs                #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE NoImplicitPrelude    #-}
+{-# LANGUAGE OverloadedLabels     #-}
+{-# LANGUAGE PolyKinds            #-}
+{-# LANGUAGE RankNTypes           #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Language.Plutus.Contract.Request where
+
+import qualified Control.Applicative                as Applicative
+import qualified Data.Aeson                         as Aeson
+import           Data.Row
+import           Data.Row.Internal                  (Subset, Unconstrained1)
+import qualified Data.Row.Records                   as Records
+import qualified Data.Row.Variants                  as Variants
+import           Language.Plutus.Contract.Events    (Event (..), Hooks (..), generalise)
+import           Language.Plutus.Contract.Resumable
+import qualified Language.Plutus.Contract.Util      as Util
+
+import           Prelude                            hiding (return, (>>=))
+import qualified Prelude
+
+-- | @Contract ρ σ a@ is a contract that expects input events of type @ρ@ and produces
+--   requests (describing the acceptable input) of type @σ@. The two type parameters
+--   are 'Data.Row.Row' rows
+--
+type Contract (ρ :: Row *) (σ :: Row *) a = Resumable (Step (Maybe (Event ρ)) (Hooks σ)) a
+
+mkRequest
+  :: forall sReq sResp req resp a.
+     ( KnownSymbol sReq
+     , KnownSymbol sResp
+     )
+    => req
+    -> (resp -> Maybe a)
+    -> Contract (sResp .== resp) (sReq .== req) a
```

Comment:
I still find it slightly weird that the `resp -> Maybe a` is built into this system. it seems like we should be able to build that functionality on top of a simpler request-response primitive, e.g.
```
mkRequest :: req -> Contract (sresp .== resp) (sReq .== req) resp
```
and then the current function could be something like `loopM (\req -> mkRequest req >>= maybe (Left req) Right) req`.

Reply:
Good point, I think can get rid of the `Maybe` now.

Reply:
I changed it to `request`, which does what you suggested, and added a `requestMaybe`